cdc/scheduler/internal/tp/agent.go

```diff
@@ -27,31 +40,589 @@ type agent struct {
+		for _, item := range batch {
+			task := item.(*dispatchTableTask)
+			if task.Epoch != a.epoch {
+				log.Info("dispatch request epoch does not match",
+					zap.String("capture", a.captureID),
+					zap.String("namespace", a.changeFeedID.Namespace),
+					zap.String("changefeed", a.changeFeedID.ID),
+					zap.Any("epoch", task.Epoch),
+					zap.Any("expected", a.epoch))
+				continue
+			}
+			if _, ok := a.runningTasks[task.TableID]; ok {
+				log.Panic("duplicate dispatch table request",
+					zap.String("capture", a.captureID),
+					zap.String("namespace", a.changeFeedID.Namespace),
+					zap.String("changefeed", a.changeFeedID.ID),
+					zap.Any("task", task))
+			}
+			a.runningTasks[task.TableID] = task
+		}
+	}
+}
+
+func (a *agent) handleDispatchTableTasks(
+	ctx context.Context,
+) (result []*schedulepb.Message, err error) {
+	a.fetchPendingTasks()
+	result = make([]*schedulepb.Message, 0)
+	for _, task := range a.runningTasks {
+		var response *schedulepb.Message
+		if task.IsRemove {
+			response = a.handleRemoveTableTask(ctx, task)
+		} else {
+			response, err = a.handleAddTableTask(ctx, task)
+		}
+		if err != nil {
+			return result, errors.Trace(err)
+		}
+		if response != nil {
+			result = append(result, response)
+		}
+	}
+	return result, nil
+}
+
+// GetLastSentCheckpointTs implement agent interface
 func (a *agent) GetLastSentCheckpointTs() (checkpointTs model.Ts) {
+	// no need to implement this.
 	return internal.CheckpointCannotProceed
 }
 
+// Close implement agent interface
 func (a *agent) Close() error {
-	return nil
+	log.Debug("agent: closing",
+		zap.String("capture", a.captureID),
+		zap.String("namespace", a.changeFeedID.Namespace),
+		zap.String("changefeed", a.changeFeedID.ID))
+	return a.trans.Close()
 }
 
-func (a *agent) handleMessage(msg []*schedulepb.Message) {
-	// s.handleMessageHeartbeat()
-	// s.handleMessageDispatchTableRequest()
+func (a *agent) newMessageHeader() *schedulepb.Message_Header {
+	return &schedulepb.Message_Header{
+		Version:        a.version,
+		OwnerRevision:  a.ownerInfo.revision,
+		ProcessorEpoch: a.epoch,
+	}
 }
 
-func (a *agent) handleMessageHeartbeat(msg *schedulepb.Heartbeat) {
-	// TODO: build s.tables from Heartbeat message.
+// handleOwnerInfo return false, if the given owner's info is staled.
+// update owner's info to the latest otherwise.
+// id: the incoming owner's capture ID
+// revision: the incoming owner's revision as generated by Etcd election.
+// version: the incoming owner's semantic version string
+func (a *agent) handleOwnerInfo(id model.CaptureID, revision int64, version string) bool {
+	if a.ownerInfo.revision.Revision == revision {
+		if a.ownerInfo.captureID != id {
+			// This panic will happen only if two messages have been received
+			// with the same ownerRev but with different ownerIDs.
+			// This should never happen unless the election via Etcd is buggy.
+			log.Panic("owner IDs do not match",
```

Comment:
This does not panic, because message from different owner should have different revision.

This is for two owner with `the same revision, but different owner ID`, which is unrecoveriable etcd error.